File: using-python/command_pattern.py
from abc import ABC, abstractmethod

# Interface for FileSystemReceiver
class FileSystemReceiver(ABC):
    @abstractmethod
    def open_file(self):
        pass

    @abstractmethod
    def write_file(self):
        pass

    @abstractmethod
    def close_file(self):
        pass

# Concrete implementations for Unix and Windows FileSystemReceiver
class UnixFileSystemReceiver(FileSystemReceiver):
    def open_file(self):
        print("Opening file in Unix OS")

    def write_file(self):
        print("Writing file in Unix OS")

    def close_file(self):
        print("Closing file in Unix OS")

class WindowsFileSystemReceiver(FileSystemReceiver):
    def open_file(self):
        print("Opening file in Windows OS")

    def write_file(self):
        print("Writing file in Windows OS")

    def close_file(self):
        print("Closing file in Windows OS")

# Interface for Command
class Command(ABC):
    @abstractmethod
    def execute(self):
        pass

# Concrete Command implementations
class OpenFileCommand(Command):
    def __init__(self, file_system):
        self.file_system = file_system

    def execute(self):
        self.file_system.open_file()

class WriteFileCommand(Command):
    def __init__(self, file_system):
        self.file_system = file_system

    def execute(self):
        self.file_system.write_file()

class CloseFileCommand(Command):
    def __init__(self, file_system):
        self.file_system = file_system

    def execute(self):
        self.file_system.close_file()

# Invoker class
class FileInvoker:
    def __init__(self, command):
        self.command = command

    def execute(self):
        self.command.execute()

# Utility class to get underlying FileSystemReceiver based on OS
class FileSystemReceiverUtil:
    @staticmethod
    def get_underlying_file_system():
        os_name = "Windows"  # For simplicity, hardcoded for demonstration
        print("Underlying OS is:", os_name)
        if os_name == "Windows":
            return WindowsFileSystemReceiver()
        else:
            return UnixFileSystemReceiver()

# Client code
def main():
    # Creating the receiver object
    fs = FileSystemReceiverUtil.get_underlying_file_system()

    # Creating command and associating with receiver
    open_file_command = OpenFileCommand(fs)

    # Creating invoker and associating with Command
    file = FileInvoker(open_file_command)

    # Perform action on invoker object
    file.execute()

    write_file_command = WriteFileCommand(fs)
    file = FileInvoker(write_file_command)
    file.execute()

    close_file_command = CloseFileCommand(fs)
    file = FileInvoker(close_file_command)
    file.execute()

if __name__ == "__main__":
    main()

File: using-typescript/command-pattern.ts
interface FileSystemReceiver {
  openFile(): void
  writeFile(): void
  closeFile(): void
}

class UnixFileSystemReceiver implements FileSystemReceiver {
  openFile(): void {
    console.log("Opening file in Unix OS")
  }

  writeFile(): void {
    console.log("Writing file in Unix OS")
  }

  closeFile(): void {
    console.log("Closing file in Unix OS")
  }
}

class WindowsFileSystemReceiver implements FileSystemReceiver {
  openFile(): void {
    console.log("Opening file in Windows OS")
  }

  writeFile(): void {
    console.log("Writing file in Windows OS")
  }

  closeFile(): void {
    console.log("Closing file in Windows OS")
  }
}

interface Command {
  execute(): void
}

class OpenFileCommand implements Command {
  private fileSystem: FileSystemReceiver

  constructor(fs: FileSystemReceiver) {
    this.fileSystem = fs
  }

  execute(): void {
    this.fileSystem.openFile()
  }
}

class WriteFileCommand implements Command {
  private fileSystem: FileSystemReceiver

  constructor(fs: FileSystemReceiver) {
    this.fileSystem = fs
  }

  execute(): void {
    this.fileSystem.writeFile()
  }
}

class CloseFileCommand implements Command {
  private fileSystem: FileSystemReceiver

  constructor(fs: FileSystemReceiver) {
    this.fileSystem = fs
  }

  execute(): void {
    this.fileSystem.closeFile()
  }
}

class FileInvoker {
  private command: Command

  constructor(c: Command) {
    this.command = c
  }

  execute(): void {
    this.command.execute()
  }
}

class FileSystemReceiverUtil {
  static getUnderlyingFileSystem(): FileSystemReceiver {
    const osName = "Windows" // Hardcoded for demonstration
    console.log("Underlying OS is:", osName)
    if (osName === "Windows") {
      return new WindowsFileSystemReceiver()
    } else {
      return new UnixFileSystemReceiver()
    }
  }
}

function main(): void {
  // Creating the receiver object
  const fs: FileSystemReceiver =
    FileSystemReceiverUtil.getUnderlyingFileSystem()

  // Creating command and associating with receiver
  const openFileCommand: Command = new OpenFileCommand(fs)

  // Creating invoker and associating with Command
  const file: FileInvoker = new FileInvoker(openFileCommand)

  // Perform action on invoker object
  file.execute()

  const writeFileCommand: Command = new WriteFileCommand(fs)
  const file2: FileInvoker = new FileInvoker(writeFileCommand)
  file2.execute()

  const closeFileCommand: Command = new CloseFileCommand(fs)
  const file3: FileInvoker = new FileInvoker(closeFileCommand)
  file3.execute()
}

// Usage
main()
